types/wlr_input_method_v2.c

```diff
@@ -101,10 +119,177 @@ static void im_get_input_popup_surface(struct wl_client *client,
 	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
 }
 
+static void im_keyboard_grab_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab) {
+	im_keyboard_grab->input_method->im_keyboard_grab = NULL;
+	if (im_keyboard_grab->keyboard) {
+		wl_list_remove(&im_keyboard_grab->keymap_listener.link);
+		wl_list_remove(&im_keyboard_grab->repeat_info_listener.link);
+
+		// send modifier state to original client
+		wlr_seat_keyboard_notify_modifiers(im_keyboard_grab->input_method->seat,
+			&im_keyboard_grab->keyboard->modifiers);
+	}
+	free(im_keyboard_grab);
+}
+
+static void im_keyboard_grab_resource_destroy(struct wl_resource *resource) {
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		im_keyboard_grab_from_resource(resource);
+	if (!im_keyboard_grab) {
+		return;
+	}
+	im_keyboard_grab_destroy(im_keyboard_grab);
+}
+
+static void im_keyboard_grab_release(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zwp_input_method_keyboard_grab_v2_interface im_keyboard_grab_impl = {
+	.release = im_keyboard_grab_release,
+};
+
+void wlr_input_method_keyboard_grab_v2_send_key(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		uint32_t time, uint32_t key, uint32_t state) {
+	zwp_input_method_keyboard_grab_v2_send_key(
+		im_keyboard_grab->resource,
+		wlr_seat_client_next_serial(im_keyboard_grab->seat_client),
+		time, key, state);
+}
+
+void wlr_input_method_keyboard_grab_v2_send_modifiers(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		struct wlr_keyboard_modifiers *modifiers) {
+	zwp_input_method_keyboard_grab_v2_send_modifiers(
+		im_keyboard_grab->resource,
+		wlr_seat_client_next_serial(im_keyboard_grab->seat_client),
+		modifiers->depressed, modifiers->latched,
+		modifiers->locked, modifiers->group);
+}
+
+static bool im_keyboard_grab_send_keymap(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	int keymap_fd = allocate_shm_file(keyboard->keymap_size);
+	if (keymap_fd < 0) {
+		wlr_log(WLR_ERROR, "creating a keymap file for %zu bytes failed",
+			keyboard->keymap_size);
+		return false;
+	}
+
+	void *ptr = mmap(NULL, keyboard->keymap_size, PROT_READ | PROT_WRITE,
+		MAP_SHARED, keymap_fd, 0);
+	if (ptr == MAP_FAILED) {
+		wlr_log(WLR_ERROR, "failed to mmap() %zu bytes",
+			keyboard->keymap_size);
+		close(keymap_fd);
+		return false;
+	}
+
+	strcpy(ptr, keyboard->keymap_string);
```

Comment:
I'd prefer to use `memcpy` here, since it has a size argument.

Reply:
Done.